### PR TITLE
Port LIS2MDL driver on top of stmemsc i/f

### PR DESCRIPTION
**Pull request by avisconti**
Port the lis2mdl sensor driver on top of the lis2mdl_StdC HAL interface (in modules/hal/st/sensor/stmemsc/). The new driver is using I2C/SPI bus.

This PR is also modifing x_nucleo_iks01a3 board and sample for a proper testing of the
driver in I2C mode.

### CODE REVIEW

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks!

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 1]
```diff
@@ -1,9 +1,11 @@
-/*
- * Copyright (c) 2018 STMicroelectronics
+/* ST Microelectronics LIS2MDL 3-axis magnetometer sensor
  *
- * LIS2MDL mag driver
+ * Copyright (c) 2019 STMicroelectronics
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: Change the year to 2018-2019

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 2]
```diff
@@ -134,7 +94,7 @@ static void lis2mdl_channel_get_mag(struct device *dev,
 	}
 
 	for (i = ofs_start; i <= ofs_stop; i++) {
-		cval = lis2mdl->mag[i] * lis2mdl->mag_fs_sensitivity;
+		cval = lis2mdl->mag[i] * 1500;
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: Is this change really related to stmemsc i/f?

File: drivers/sensor/lis2mdl/lis2mdl.c [hunk 3]
```diff
@@ -148,7 +108,7 @@ static void lis2mdl_channel_get_temp(struct device *dev,
 	struct lis2mdl_data *drv_data = dev->driver_data;
 
 	val->val1 = drv_data->temp_sample / 100;
-	val->val2 = drv_data->temp_sample % 100;
+	val->val2 = (drv_data->temp_sample % 100) * 10000;
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.c]: Is this change really related to stmemsc i/f?

File: drivers/sensor/lis2mdl/lis2mdl.h [hunk 4]
```diff
@@ -1,114 +1,54 @@
-/*
- * Copyright (c) 2018 STMicroelectronics
+/* ST Microelectronics LIS2MDL 3-axis magnetometer sensor
  *
- * LIS2MDL mag header file
+ * Copyright (c) 2019 STMicroelectronics
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.h]: Change the year to 2018-2019

File: drivers/sensor/lis2mdl/lis2mdl.h [hunk 5]
```diff
@@ -1,114 +1,54 @@
...
+ * Datasheet:
+ * https://www.st.com/resource/en/datasheet/lis2mdl.pdf
  */
 
 #ifndef __MAG_LIS2MDL_H
 #define __MAG_LIS2MDL_H
 
+#include <drivers/spi.h>
 #include <drivers/gpio.h>
+#include <sensor.h>
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl.h]: drivers/sensor.h

File: drivers/sensor/lis2mdl/lis2mdl_trigger.c [hunk 6]
```diff
@@ -1,9 +1,11 @@
-/*
- * Copyright (c) 2018 STMicroelectronics
+/* ST Microelectronics LIS2MDL 3-axis magnetometer sensor
  *
- * LIS2MDL mag interrupt configuration and management
+ * Copyright (c) 2019 STMicroelectronics
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl_trigger.c]: Change the year to 2018-2019

File: drivers/sensor/lis2mdl/lis2mdl_trigger.c [hunk 7]
```diff
@@ -155,7 +135,18 @@ int lis2mdl_init_interrupt(struct device *dev)
 	lis2mdl->dev = dev;
 #endif
 
-	gpio_pin_enable_callback(lis2mdl->gpio, config->gpio_pin);
+	gpio_pin_configure(lis2mdl->gpio, config->gpio_pin,
```
Comment (MaureenHelm) [inline drivers/sensor/lis2mdl/lis2mdl_trigger.c]: Is this change really related to stmemsc i/f?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/lis2mdl/lis2mdl_trigger.c (hunk 7 above)
Comment (avisconti) [reply drivers/sensor/lis2mdl/lis2mdl_trigger.c]: Not related to stmemsc. I thought it was good to re-write it in the way all the other drivers, at least the ST ones, are done. Moreover, I think that it is better to create the callback thread prior to enable the callback.

File: drivers/sensor/lis2mdl/lis2mdl.c (hunk 3 above)
Comment (avisconti) [reply drivers/sensor/lis2mdl/lis2mdl.c]: No, I realized this was a bug. I can put it in a separate commit

File: drivers/sensor/lis2mdl/lis2mdl_trigger.c (hunk 7 above)
Comment (avisconti) [reply drivers/sensor/lis2mdl/lis2mdl_trigger.c]: I would keep it in the same commit even if not strictly related to stmemsc. Do you agree?

File: drivers/sensor/lis2mdl/lis2mdl.c (hunk 2 above)
Comment (avisconti) [reply drivers/sensor/lis2mdl/lis2mdl.c]: No, it is a code cleaning. The sensitivity does not change and it is always 1500 mgauss/LSB.
I would keep it in the same commit even if not strictly related to stmemsc. Do you agree?

File: drivers/sensor/lis2mdl/lis2mdl_trigger.c (hunk 7 above)
Comment (MaureenHelm) [reply drivers/sensor/lis2mdl/lis2mdl_trigger.c]: It's fine to keep in the same PR, but I'd prefer it to be in a separate commit if it's not related to stmemsc

**Comment by avisconti** [comment]
@MaureenHelm 
I re-pushed it adding your changes except a couple of things highlighted above. Please let me know if you agree.

Moreover, I noticed that I made a type concerning CS_GPIOS DT instance, found just because I built it for SPI as well. So I also made that change.

File: drivers/sensor/lis2mdl/lis2mdl_trigger.c (hunk 7 above)
Comment (avisconti) [reply drivers/sensor/lis2mdl/lis2mdl_trigger.c]: ACK.
Done in this way and re-pushed.

File: drivers/sensor/lis2mdl/lis2mdl.c (hunk 2 above)
Comment (MaureenHelm) [reply drivers/sensor/lis2mdl/lis2mdl.c]: No, please move to a separate commit

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/lis2mdl/lis2mdl.c (hunk 2 above)
Comment (avisconti) [reply drivers/sensor/lis2mdl/lis2mdl.c]: ok, done!

**Comment by avisconti** [comment]
@MaureenHelm 
Every point should be addressed now.

**Review by MaureenHelm — APPROVED** [review]
